File: common/src/rc.c
#include "common/rc.h"
#include <string.h>

static char const name[9][9] = {"done",   "end", "next",   "notfound", "efail",
                                "einval", "eio", "enomem", "eparse"};

enum rc rc_from_str(unsigned len, char const *str, enum rc *rc)
{
    for (unsigned i = 0; i < 9; ++i)
    {
        if (!strncmp(name[i], str, len))
        {
            *rc = (enum rc)i;
            return RC_DONE;
        }
    }
    return RC_EINVAL;
}

File: rest/deciphon_rest/rc.py
from enum import Enum, IntEnum
from pydantic import BaseModel
from .logger import logger


class RC(IntEnum):
    RC_DONE = 0
    RC_END = 1
    RC_NEXT = 2
    RC_NOTFOUND = 3
    RC_EFAIL = 4
    RC_EINVAL = 5
    RC_EIO = 6
    RC_ENOMEM = 7
    RC_EPARSE = 8


class ReturnCode(str, Enum):
    RC_DONE = "done"
    RC_END = "end"
    RC_NEXT = "next"
    RC_NOTFOUND = "notfound"
    RC_EFAIL = "efail"
    RC_EINVAL = "einval"
    RC_EIO = "eio"
    RC_ENOMEM = "enomem"
    RC_EPARSE = "eparse"

    class Config:
        use_enum_values = True


class ReturnData(BaseModel):
    rc: ReturnCode = ReturnCode.RC_DONE
    error: str = ""


def return_data(rc):
    return ReturnData(rc=ReturnCode[RC(rc).name], error=logger.pop())
